File: server/Socks/Protocols/Packet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Glove;

namespace SockScape {
    class Packet {
        // (squid)
        public static readonly byte[] MagicNumber = { 0xF0, 0x9F, 0xA6, 0x91 };

        public static Packet FromBytes(byte[] raw) {
            if(raw.Length < 7)
                return null;

            Packet packet = new Packet();

            if(!raw.Subset(0, 4).SequenceEqual(MagicNumber))
                return null;

            packet.Id = raw[4];
            var regionCount = raw[5];
            var regionLengths = new List<uint>();
            var headerPtr = 6;
            for(var i = 0; i < regionCount; ++i) {
                regionLengths.Add(0);
                var first = raw[headerPtr];
                if(first < 254) {
                    regionLengths[i] = first;
                    ++headerPtr;
                } else if(first == 254) {
                    if(headerPtr + 3 < raw.Length)
                        return null;
                    regionLengths[i] = raw.Subset(headerPtr + 1, 2).UnpackUInt16();
                    headerPtr += 3;
                } else {
                    if(headerPtr + 5 < raw.Length)
                        return null;
                    regionLengths[i] = raw.Subset(headerPtr + 1, 4).UnpackUInt32();
                    headerPtr += 5;
                }

                if(headerPtr > raw.Length)
                    return null;
            }

            if(headerPtr + regionLengths.Sum(x => x) > raw.Length)
                return null;
             
            long bodyPtr = headerPtr;
            foreach(var regionLength in regionLengths) {
                // FLAG TODO this could fail if one region exceeds 2^31-1 in size, check later
                packet.Regions.Add(raw.Subset((int)bodyPtr, (int)regionLength));
                bodyPtr += regionLength;
            }

            return packet;
        }
        
        private readonly List<byte[]> Regions = new List<byte[]>();
        public int Id { get; private set; }
        public int RegionCount
            => Regions.Count;

        protected Packet() { }

        public Packet(Enum id, params object[] regions) {
            Initialize((int)Convert.ChangeType(id, id.GetTypeCode()), regions);
        }

        public Packet(int id, params object[] regions) {
            Initialize(id, regions);
        }

        private void Initialize(int id, object[] regions) {
            Id = id;

            foreach(var region in regions)
                AddRegion(region);
        }

        public Region this[int i] 
            => new Region(Regions[i]);

        public Packet AddRegion(object region) {
            if(region.GetType() == typeof(byte[]))
                Regions.Add((byte[])region);
            else if(region.GetType() == typeof(string))
                Regions.Add(((string)region).GetBytes());
            else if(region.GetType() == typeof(byte))
                Regions.Add(new[] { (byte)region });
            else
                Console.WriteLine($"Could not add region {region} of type {region.GetType()}.");

            return this;
        }

        public Packet AddRegions(params object[] regions) {
            foreach(var region in regions)
                AddRegion(region);

            return this;
        }

        public bool CheckRegionLengths(int startIndex, params int[] lengths) {
            if(startIndex + lengths.Length > RegionCount)
                return false;

            for(int i = 0; i < lengths.Length; ++i) {
                if(this[startIndex + i].Raw.Length != lengths[i])
                    return false;
            }

            return true;
        }

        public bool CheckRegionsMaxLength(int startIndex, params int[] lengths) {
            if(startIndex + lengths.Length > RegionCount)
                return false;

            for(int i = 0; i < lengths.Length; ++i) {
                if(this[startIndex + i].Raw.Length > lengths[i])
                    return false;
            }

            return true;
        }

        public bool CheckAllMaxLength(int length, int startIndex = 0) {
            if(startIndex > RegionCount)
                return false;

            for(int i = startIndex; i < RegionCount; ++i) {
                if(this[i].Raw.Length > length)
                    return false;
            }

            return true;
        }

        public byte[] GetBytes() {
            var header = new List<byte>();
            header.AddRange(MagicNumber);
            header.Add((byte)Id);
            header.Add((byte)RegionCount);
            
            IEnumerable<byte> body = new byte[0];
            foreach(var region in Regions) {
                if(region.Length < 0xFE)
                    header.Add((byte)region.Length);
                else if(region.Length <= 0xFFFF) {
                    header.Add(0xFE);
                    header.AddRange(((UInt16)region.Length).Pack());
                } else {
                    header.Add(0xFF);
                    header.AddRange(region.Length.Pack());
                }

                body = body.Concat(region);
            }

            return header.Concat(body).ToArray();
        }

        public class Region {
            private byte[] Data { get; }

            public Region(byte[] data) {
                Data = data;
            }

            public static implicit operator byte[] (Region region) => region.Data;
            public byte[] Raw
                => this;

            public static implicit operator string(Region region) {
                try {
                    return Encoding.UTF8.GetString(region.Data);
                } catch {
                    return Encoding.ASCII.GetString(region.Data);
                }
            }
            public string Str
                => this;
        }
    }

    // FLAG TODO determine if you still need this thing, probably not
    public class PacketBuffer {
        private readonly short MaxSize;
        private Dictionary<uint, byte[]> Buffer
            = new Dictionary<uint, byte[]>();

        public PacketBuffer(short maxSize = 10) {
            MaxSize = maxSize;
        }

        public void Add(uint id, byte[] packet) {
            Buffer[id] = packet;

            if(Buffer.Count > MaxSize)
                Buffer =
                    Buffer.Where(x => x.Key >= id - MaxSize)
                        .ToDictionary(x => x.Key, x => x.Value);
        }

        public byte[] this[uint i]
            => Buffer[i];

        public bool HasId(uint id)
            => Buffer.ContainsKey(id);

        public void Clear()
            => Buffer.Clear();
    }
}

File: client/src/sock/Packet.ts
const enum kMasterId {
    KeyExchange = 1,
    LoginAttempt,
    RegistrationAttempt,
    ServerListing
}

const enum kSlaveId {
    
}

class Packet {
    private static magicNumber: Uint8Array = new Uint8Array([0xF0, 0x9F, 0xA6, 0x91]);

    private _id: number;
    public get id(): number {
        return this._id;
    }

    private _regions: Uint8Array[] = [];
    public get regions(): Uint8Array[] {
        return this._regions;
    }
    public get regionCount(): number {
        return this._regions.length;
    }
    public getRegion(region: number): Uint8Array {
        return this._regions[region];
    }
    public getRegionString(region: number): string {
        return this.getRegion(region).toString();
    }
    public addRegion(region: object): Packet {
        if(typeof region == "string")
            this._regions.push((<string>region).toByteArray());
        else if(region instanceof Uint8Array)
            this._regions.push(region);

        this[this.regionCount-1] = this._regions[this.regionCount-1];
        return this;
    }

    private constructor() {}

    public static create(id: number, regions: any[]): Packet {
        var packet = new Packet;
        packet._id = id;
        regions.forEach(region => {
            packet.addRegion(region);
        });

        return packet;
    }

    public static fromBytes(bytes: Uint8Array): Packet {
        var packet = new Packet;

        if(!bytes.subarray(0, 4).every((v, i) => v === Packet.magicNumber[i]))
            return null;

        packet._id = bytes[4];
        var regionCount = bytes[5];
        var regionLengths = [];
        var ptr = 6;
        for(var i = 0; i < regionCount; ++i) {
            if(bytes[ptr] < 0xFE)
                regionLengths.push(bytes[ptr]);
            else if(bytes[ptr] == 0xFE) {
                regionLengths.push(bytes.unpackUint16(ptr + 1));
                ptr += 2;
            } else {
                regionLengths.push(bytes.unpackUint32(ptr + 1));
                ptr += 4;
            }

            ++ptr;
        }

        for(var i = 0; i < regionCount; ++i) {
            packet.addRegion(bytes.subarray(ptr, ptr + regionLengths[i]));
            ptr += regionLengths[i];
        }

        return packet;
    }

    public getBytes(): Uint8Array {
        var headerSize = 6, bodySize = 0;
        this._regions.forEach(region => {
            bodySize += region.byteLength;

            ++headerSize;
            if(region.byteLength >= 0xFE && region.byteLength <= 0xFFFF)
                headerSize += 2;
            else if(region.byteLength > 0xFFFF)
                headerSize += 4;
        });

        var buffer = new Uint8Array(headerSize + bodySize);
        var headerPtr = 6, bodyPtr = headerSize;
        buffer.set(Packet.magicNumber, 0);
        buffer[4] = this._id % 256;
        buffer[5] = this._regions.length;
        this._regions.forEach(region => {
            var regionLength = region.byteLength;
            if(regionLength < 0xFE)
                buffer[headerPtr] = regionLength;
            else if(regionLength >= 0xFE && regionLength <= 0xFFFF) {
                buffer[headerPtr] = 0xFE;
                buffer.set(regionLength.packUint16(), headerPtr + 1);
                headerPtr += 2;
            } else {
                buffer[headerPtr] = 0xFF;
                buffer.set(regionLength.packUint32(), headerPtr + 1);
                headerPtr += 4;
            }
            ++headerPtr;

            buffer.set(region, bodyPtr);
            bodyPtr += regionLength;
        });
        
        return buffer;
    }
}